xcpmd/src/screen-module.c

```diff
@@ -0,0 +1,320 @@
+#include "rules.h"
+#include <math.h>
+
+//Function prototypes
+void screen_on(struct arg_node * args);
+void screen_off(struct arg_node * args);
+void set_backlight(struct arg_node * args);
+void increase_backlight(struct arg_node * args);
+void decrease_backlight(struct arg_node * args);
+static int get_brightness(void);
+static int get_raw_brightness(int *brightness, int *max_brightness);
+static void set_brightness(int percentage);
+
+
+//Private data structures
+struct action_table_row {
+    char * name;
+    void (* func)(struct arg_node *);
+    char * prototype;
+    char * pretty_prototype;
+};
+
+
+//Private data
+static struct action_table_row action_table[] = {
+    {"screenOn"          , screen_on          , "n" , "void"                    } ,
+    {"screenOff"         , screen_off         , "n" , "void"                    } ,
+    {"setBacklight"      , set_backlight      , "i" , "int backlight_percent"   } ,
+    {"increaseBacklight" , increase_backlight , "i" , "int percent_to_increase" } ,
+    {"decreaseBacklight" , decrease_backlight , "i" , "int percent_to_decrease" }
+};
+
+static unsigned int num_action_types = sizeof(action_table) / sizeof(action_table[0]);
+static int times_loaded = 0;
+
+
+//Registers this module's action types.
+//The constructor attribute causes this function to run at load (dlopen()) time.
+__attribute__ ((constructor)) static void init_module() {
+
+    unsigned int i;
+
+    if (times_loaded > 0)
+        return;
+
+    for (i=0; i < num_action_types; ++i) {
+        add_action_type(action_table[i].name, action_table[i].func, action_table[i].prototype, action_table[i].pretty_prototype);
+    }
+}
+
+
+//Cleans up after this module.
+//The destructor attribute causes this to run at unload (dlclose()) time.
+__attribute__ ((destructor)) static void uninit_module() {
+
+    --times_loaded;
+
+    //if (times_loaded > 0)
+    //    return;
+
+    return;
+}
+
+
+//Actions
+//Turns all screens on.
+void screen_on(struct arg_node * args) {
+    com_citrix_xenclient_surfman_dpms_on_(xcdbus_conn, SURFMAN_SERVICE, SURFMAN_PATH);
+}
+
+
+//Turns all screens off.
+void screen_off(struct arg_node * args) {
+    com_citrix_xenclient_surfman_dpms_off_(xcdbus_conn, SURFMAN_SERVICE, SURFMAN_PATH);
+}
+
+
+//Sets the backlight to as close to the desired percentage as surfman allows.
+void set_backlight(struct arg_node * args) {
+
+    struct arg_node * node = get_arg(args, 0);
+    set_brightness(node->arg.i);
+}
+
+
+//Increases the backlight by as close to the current percentage as surfman allows.
+void increase_backlight(struct arg_node * args) {
+
+    struct arg_node * node = get_arg(args, 0);
+    int old_backlight, new_backlight;
+
+    old_backlight = get_brightness();
+    if (old_backlight < 0) {
+        return;
+    }
+
+    new_backlight = old_backlight + node->arg.i;
+
+    if (new_backlight > 100) {
+        new_backlight = 100;
+    }
+    else if (new_backlight < old_backlight) {
```

Comment:
Is this to handle integer overflow, in case the increment is negative or bigger than `INT_MAX - 100`?
This looks weird, maybe testing first that `0 <= increment <= 100` would make more sense...